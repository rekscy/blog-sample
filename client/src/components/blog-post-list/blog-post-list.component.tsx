import { DateTime } from "luxon";
import React from "react";
import { useHistory } from "react-router";
import { BlogPostListProps } from "./blog-post-list.types";
import "./blog-post-list.style.css";

const BlogPostListComponent = (props: BlogPostListProps) => {
  const { blogPosts } = props;
  const history = useHistory();

  const redirectToBlogPost = (id: string) => {
    history.push(`/blog/${id}`);
  };

  return (
    <div className="blog-post-list">
      <h2>Last posts</h2>
      {blogPosts.map(({ id, title, nbrComments, content, createdAt }) => {
        return (
          <div
            data-testid="blog-post-container"
            className="blog-post-container"
            key={id}
            onClick={() => redirectToBlogPost(id)}
          >
            <div className="blog-post-header">
              <h3>{title}</h3>
            </div>
            <div className="blog-post-body">
              <p className="blog-post-content">{content}</p>
            </div>
            <div className="blog-post-footer">
              <div className="blog-post-nbr-comments">
                <i className="fas fa-comments" />
                <span>{nbrComments}</span>
              </div>
              <div className="blog-post-date">
                <i className="fas fa-calendar-alt" />
                <span>
                  {DateTime.fromMillis(createdAt)
                    .setLocale("FR-ch")
                    .toLocaleString(DateTime.DATETIME_SHORT)}
                </span>
              </div>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default BlogPostListComponent;
