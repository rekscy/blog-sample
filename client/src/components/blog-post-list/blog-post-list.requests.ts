import axios from "axios";
import { config } from "../../config/config";
import { BlogPosts } from "../blog-post/blog-post.types";

/**
 * Get blog posts from server.
 */
export const fetchBlogPosts = async () => {
  const response = await axios.get<BlogPosts>("/blog-posts", {
    baseURL: config.backendApiBaseEndpoint,
  });
  return response.data;
};
