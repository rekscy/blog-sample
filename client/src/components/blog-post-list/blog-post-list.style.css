.blog-post-list {
  padding: 0 1rem;
}

.blog-post-list > h2{
  text-transform: uppercase;
  text-align: left;
  color: #5629c8;
  font-size: 2rem;
  border-bottom: 1px solid #e5dcff;
}

.blog-post-container {
  text-align: left;
  color: #373737;
  font-weight: normal;
  font-size: 1rem;
  margin: 0;
  padding: 1rem;
}

.blog-post-container:hover {
  background: #eeeeee;
  cursor: pointer;
}

.blog-post-header > h3 {
  text-align: left;
  color: #946fff;
  font-size: 1.5rem;
}

.blog-post-container:not(:last-child) {
  border-bottom: 1px solid #d6d6d6;
}

.blog-post-footer {
  display: flex;
  justify-content: space-between;
}

.blog-post-nbr-comments {
  font-weight: bold;
}

.blog-post-nbr-comments > span {
  margin-left: 10px;
}
.blog-post-date {
  justify-content: flex-end;
}

.blog-post-date > span {
  margin-left: 10px;
}
