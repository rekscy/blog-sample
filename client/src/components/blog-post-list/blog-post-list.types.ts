import { BlogPosts } from "../blog-post/blog-post.types";

export interface BlogPostListProps {
  blogPosts: BlogPosts;
}
