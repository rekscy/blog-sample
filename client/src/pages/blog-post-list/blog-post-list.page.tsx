import React, { useEffect, useState } from "react";
import { store } from "react-notifications-component";
import { createNotificationFromAxiosHttpError } from "../../utils/global/global.utils";
import BlogPostListComponent from "../../components/blog-post-list/blog-post-list.component";
import Loader from "../../components/loader/loader.component";
import { BlogPosts } from "../../components/blog-post/blog-post.types";
import { fetchBlogPosts } from "../../components/blog-post-list/blog-post-list.requests";

const BlogPostListPage = () => {
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [blogPostsList, setBlogPostsList] = useState<BlogPosts>([]);

  useEffect(() => {
    const fetch = async () => {
      const response = await fetchBlogPosts();
      setIsLoading(false);
      setBlogPostsList(response);
    };

    fetch().catch((e) => {
      const notification = createNotificationFromAxiosHttpError(e);
      store.addNotification(notification);
    });
  }, []);

  if (isLoading) {
    return <Loader />;
  }

  return <BlogPostListComponent blogPosts={blogPostsList} />;
};

export default BlogPostListPage;
