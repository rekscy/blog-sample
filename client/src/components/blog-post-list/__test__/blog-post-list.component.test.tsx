import { shallow } from "enzyme";
import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Router } from "react-router";
import { createMemoryHistory } from "history";
import { postsMockedData } from "./blog-post-list.mocked-data";
import BlogPostListComponent from "./blog-post-list.component";

describe("Blog post list Component", () => {
  it("expect to render Blog post Component correctly", () => {
    expect(
      shallow(<BlogPostListComponent blogPosts={postsMockedData} />)
    ).toMatchSnapshot();
  });
  it("should go to blog post after click", async () => {
    const history = createMemoryHistory();

    render(
      <Router history={history}>
        <BlogPostListComponent blogPosts={postsMockedData} />
      </Router>
    );
    const item = screen.getByTestId("blog-post-container");
    expect(item).toBeInTheDocument();
    userEvent.click(item);
    expect(history.length).toBe(2);
    expect(history.location.pathname).toBe("/blog/abc123");
  });
});
