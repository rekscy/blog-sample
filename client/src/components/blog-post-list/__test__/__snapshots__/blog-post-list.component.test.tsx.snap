// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog post list Component expect to render Blog post Component correctly 1`] = `ShallowWrapper {}`;
